File: utilities/aggregate_over_traces.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <float.h>
#include <assert.h>

#include "../src/trace.h"

void usage()
{
    fprintf(stderr, "Usage: ./aggregate_over_traces (min|max|sum) output.bin.trace file(s).bin.trace\n");
}

int 
main( int argc, char** argv )
{
    if( argc <= 3 )
    {
        usage();
        exit( -1 );
    }

    /* determine the aggregate type */
    TRACE_TYPE (*agg_fn)( const TRACE_TYPE, const TRACE_TYPE ) = NULL;

    if( 0 == strcmp("max", argv[1] ) )
    {
        agg_fn = trace_max_agg;
    } else if( 0 == strcmp("min", argv[1] ) ) {
        agg_fn = trace_min_agg;
    } else if( 0 == strcmp("sum", argv[1] ) ) {
        agg_fn = trace_sum_agg;
    } else {
        fprintf( stderr, "FATAL     : Unrecognized aggregate '%s'\n", argv[1] );
        exit( -1 );
    }

    /* load the first trace as the trace that we will update */
    struct trace_t* update_trace;
    load_trace_from_file( &update_trace, argv[3] );
    
    int i;
    for( i = 1; i < argc-3; i++ )
    {
        struct trace_t* curr_trace;
        load_trace_from_file( &curr_trace, argv[i+3] );
        aggregate_over_trace_pairs( update_trace, curr_trace, agg_fn );
        close_traces( curr_trace );
    }
    
    write_trace_to_file( update_trace, argv[2] );
    
    close_traces( update_trace );
    
    return 0;
}

File: utilities/aggregate_over_traces.py
#!/usr/bin/python

import sys
import os

# add python_lib to sys.path
sys.path.insert(0, os.path.normpath( sys.path[0] + "/../python_lib") )
from trace import *

def usage():
    print "Usage: ./aggregate_over_traces.py (min|max|sum) output.bin.trace file(s).bin.trace"
    sys.exit(1)

def determine_aggregate_type( arg ):
    """
    Determine trace aggregate function from command line argument
    """
    if arg == "max":
        agg_fn_ptr = statmap_o.trace_agg_max
    elif arg == "min":
        agg_fn_ptr = statmap_o.trace_agg_min
    elif arg == "sum":
        agg_fn_ptr = statmap_o.trace_agg_sum
    else:
        raise Exception("Invalid aggregate type: %s" % arg )

    return agg_fn_ptr

def main():
    if len(sys.argv) <= 3: usage()

    agg_fn = determine_aggregate_type( sys.argv[1] )

    # load the first trace as the trace we will update
    update_trace = load_c_trace_from_file( sys.argv[3] )

    # loop over the .bin.trace input files
    for trace_fname in sys.argv[4:]:
        curr_trace = load_c_trace_from_file( trace_fname )
        aggregate_over_trace_pairs( update_trace, curr_trace, agg_fn )
        close_c_traces( curr_trace )

    # write final update trace to file
    write_c_trace_to_file( update_trace, sys.argv[2] )
    close_c_traces( update_trace )

if __name__ == "__main__": main()
